sources/api/thar-be-updates/src/main.rs

```diff
@@ -0,0 +1,427 @@
+    }
+    Ok(())
+}
+
+/// "Activates" the staged update by letting updog set up the appropriate boot flags
+fn activate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    debug!("Spawning 'updog update-apply'");
+    let output = Command::new("updog")
+        .arg("update-apply")
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::ActivateUpdate, &output);
+    if output.status.success() {
+        status.mark_staging_partition_next_to_boot()?;
+    } else {
+        warn!("Failed to activate the update with updog");
+    }
+    Ok(())
+}
+
+/// "Deactivates" the staged update by rolling back actions done by `activate_update`
+fn deactivate_update(status: &mut UpdateStatus<update_metadata::Update>) -> Result<()> {
+    debug!("Spawning 'updog revert-update-apply'");
+    let output = Command::new("updog")
+        .arg("revert-update-apply")
+        .output()
+        .context(error::Updog)?;
+    status.set_recent_command_info(UpdateCommand::DeactivateUpdate, &output);
+    if output.status.success() {
+        status.unmark_staging_partition_next_to_boot()?;
+    } else {
+        warn!("Failed to deactivate the update with updog");
+    }
+    Ok(())
+}
+
+fn run() -> Result<()> {
+    // Parse and store the args passed to the program
+    let args = parse_args(env::args());
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed)
+        .context(error::Logger)?;
+
+    // Open the lockfile for concurrency control, create it if it doesn't exist
+    let lockfile = File::create(UPDATE_LOCKFILE).context(error::LockFile {
+        path: UPDATE_LOCKFILE,
+    })?;
+
+    // Check if the update status file exists. If it doesn't, create and initialize it.
+    if !Path::new(UPDATE_STATUS_FILE).is_file() {
+        // Get an exclusive lock for creating the update status file
+        lock_exclusive(&lockfile)?;
+        initialize_update_status()?;
+        unlock(&lockfile)?;
+    }
+
+    match args.subcommand {
+        // 'refresh-updates' is allowed under every update state
+        UpdateCommand::RefreshUpdates => {
+            lock_exclusive(&lockfile)?;
+            let mut update_status = get_update_status()?;
+            match fork() {
+                Ok(ForkResult::Parent { child, .. }) => {
+                    debug!("forked child pid: {}", child);
+                    // Exit immediately as the parent
+                    // Parent's lockfile fd will close but child will still have a duplicate fd
+                    return Ok(());
+                }
+                Ok(ForkResult::Child) => {
+                    // 'refresh_update' is allowed under every update state
+                    let is_chosen_update_availabe =
+                        refresh_update(&mut update_status, &args.socket_path)?;
```

Comment:
It's fairly difficult to keep track of the mutations of the state file - it gets modified (twice) as a side effect of the `refresh_update` call, and also here when we observe the new state and modify it again.

My preference would be to handle it all within the "state" object, or failing that have the helper methods make  all necessary changes.

Reply:
Hopefully the refactor makes it easier to keep track!